oshdb-api/src/main/java/org/heigit/ohsome/oshdb/api/mapreducer/backend/MapReducerJdbc.java

```diff
@@ -48,29 +46,52 @@ public boolean isActive() {
     return true;
   }
 
-  protected ResultSet getOshCellsRawDataFromDb(CellIdRange cellIdRange)
+  protected List<Map.Entry<String, CellId>> getOshTableCellIds(CellIdRange cellIdRange)
       throws SQLException {
-    String sqlQuery = this.typeFilter.stream()
-        .map(osmType ->
-            TableNames.forOSMType(osmType).map(tn -> tn.toString(this.oshdb.prefix()))
-        )
-        .filter(Optional::isPresent).map(Optional::get)
-        .map(tn -> "(select data from " + tn + " where level = ?1 and id between ?2 and ?3)")
-        .collect(Collectors.joining(" union all "));
-    PreparedStatement pstmt = ((OSHDBJdbc) this.oshdb).getConnection().prepareStatement(sqlQuery);
-    pstmt.setInt(1, cellIdRange.getStart().getZoomLevel());
-    pstmt.setLong(2, cellIdRange.getStart().getId());
-    pstmt.setLong(3, cellIdRange.getEnd().getId());
-    return pstmt.executeQuery();
+    var cellIds = new ArrayList<Map.Entry<String, CellId>>();
+    for (var type : typeFilter) {
+      var table = TableNames.forOSMType(type).map(tn -> tn.toString(this.oshdb.prefix()));
+      if (table.isPresent()) {
+        var tableName = table.get();
+        var sqlQuery = String.format(
+            "(select level, id from %s where level = ?1 and id between ?2 and ?3)", tableName);
+        try (var pstmt = ((OSHDBJdbc) this.oshdb).getConnection().prepareStatement(sqlQuery)) {
+          pstmt.setInt(1, cellIdRange.getStart().getZoomLevel());
+          pstmt.setLong(2, cellIdRange.getStart().getId());
+          pstmt.setLong(3, cellIdRange.getEnd().getId());
+          try (var rs = pstmt.executeQuery()) {
+            while (rs.next()) {
+              cellIds.add(Map.entry(tableName, new CellId(rs.getInt(1), rs.getLong(2))));
+            }
+          }
+        }
+      }
+    }
+    return cellIds;
   }
 
   /**
-   * Returns data of one cell from the raw data stream.
+   * Returns data of one cell.
    */
-  protected GridOSHEntity readOshCellRawData(ResultSet oshCellsRawData)
-      throws IOException, ClassNotFoundException, SQLException {
-    return (GridOSHEntity)
-        (new ObjectInputStream(oshCellsRawData.getBinaryStream(1))).readObject();
+  protected GridOSHEntity readOneGridCell(Map.Entry<String, CellId> entry) {
+    var table = entry.getKey();
+    var cellId = entry.getValue();
+    var sqlQuery = String.format("select data from %s where level = ? and id = ?", table);
+    try (var pstmt = ((OSHDBJdbc) this.oshdb).getConnection().prepareStatement(sqlQuery)) {
```

Comment:
would it make sense to store this in a shared prepared statement, instead of recreating it for every single fetching of a single grid cell (similar to how it is implemented in the [tagtranslator](https://github.com/GIScience/oshdb/blob/master/oshdb-util/src/main/java/org/heigit/ohsome/oshdb/util/tagtranslator/TagTranslator.java#L84-L86))? or is that not possible because of concurrency issues?